handler_test.go

```diff
@@ -0,0 +1,1982 @@
+				Qtype: dns.TypeA,
+				Rcode: dns.RcodeNameError,
+			},
+		},
+		expectedStatuses: []ocpnetworkapiv1alpha1.DNSNameResolverStatus{
+			{
+				ResolvedNames: []ocpnetworkapiv1alpha1.DNSNameResolverResolvedName{
+					{
+						DNSName: "www.example.com.",
+						ResolvedAddresses: []ocpnetworkapiv1alpha1.DNSNameResolverResolvedAddress{
+							{
+								IP:         "1.1.1.1",
+								TTLSeconds: 30,
+							},
+							{
+								IP:         "1.1.1.2",
+								TTLSeconds: 30,
+							},
+						},
+						ResolutionFailures: 1,
+						Conditions: []metav1.Condition{
+							{
+								Type:    ConditionDegraded,
+								Status:  metav1.ConditionTrue,
+								Reason:  dns.RcodeToString[dns.RcodeNameError],
+								Message: rcodeMessage[dns.RcodeNameError],
+							},
+						},
+					},
+				},
+			},
+		},
+	},
+	{
+		name: "Increment resolutionFailure field in regular dns name resolver object's status and update ttlSeconds and lastLookupTime fields",
+		dnsNameResolvers: []ocpnetworkapiv1alpha1.DNSNameResolver{
+			{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "regular",
+					Namespace: "dns",
+				},
+				Spec: ocpnetworkapiv1alpha1.DNSNameResolverSpec{
+					Name: "www.example.com.",
+				},
+				Status: ocpnetworkapiv1alpha1.DNSNameResolverStatus{
+					ResolvedNames: []ocpnetworkapiv1alpha1.DNSNameResolverResolvedName{
+						{
+							DNSName: "www.example.com.",
+							ResolvedAddresses: []ocpnetworkapiv1alpha1.DNSNameResolverResolvedAddress{
+								{
+									IP:         "1.1.1.1",
+									TTLSeconds: 2,
+								},
+								{
+									IP:         "1.1.1.2",
+									TTLSeconds: 2,
+								},
+							},
+							ResolutionFailures: 1,
+							Conditions: []metav1.Condition{
+								{
+									Type:    ConditionDegraded,
+									Status:  metav1.ConditionTrue,
+									Reason:  dns.RcodeToString[dns.RcodeNameError],
+									Message: rcodeMessage[dns.RcodeNameError],
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+		dnsQueryTestCases: []test.Case{
+			{
+				Qname: "www.example.com.",
+				Qtype: dns.TypeA,
+				Rcode: dns.RcodeNameError,
+			},
+		},
+		expectedStatuses: []ocpnetworkapiv1alpha1.DNSNameResolverStatus{
+			{
+				ResolvedNames: []ocpnetworkapiv1alpha1.DNSNameResolverResolvedName{
+					{
+
+						DNSName: "www.example.com.",
+						ResolvedAddresses: []ocpnetworkapiv1alpha1.DNSNameResolverResolvedAddress{
+							{
+								IP:         "1.1.1.1",
+								TTLSeconds: 5,
+							},
+							{
+								IP:         "1.1.1.2",
+								TTLSeconds: 5,
+							},
+						},
+						ResolutionFailures: 2,
+						Conditions: []metav1.Condition{
+							{
+								Type:    ConditionDegraded,
+								Status:  metav1.ConditionTrue,
+								Reason:  dns.RcodeToString[dns.RcodeNameError],
+								Message: rcodeMessage[dns.RcodeNameError],
+							},
+						},
+					},
+				},
+			},
+		},
+	},
+	{
+		name: "Remove resolved dns name from regular dns name resolver object's status",
```

Comment:
This actually due to `ResolutionFailures` value going beyond 5.

The TTLSeconds value is chosen as 0 to depict a realistic situation where the TTL has expired triggered a DNS lookup by the DNSNameResolver controller. As the DNS name lookup has been failing continuously (`ResolutionFailures` value being 5), the plugin has updated the `TTLSeconds` to 5 for every failure (https://github.com/openshift/coredns-ocp-dnsnameresolver/pull/1#discussion_r1526205565). As this also fails and as the `ResolutionFailures` value incremented beyond 5, the complete resolved field is removed by the plugin.

The removal of the IP addresses on TTL expiration will not be done by the plugin. It will be done by the DNSNameResolver controller waiting for a grace period after the expiration of the TTL of the IP addresses.

Reply:
> This actually due to ResolutionFailures value going beyond 5.

Ah ok, let's state this in the test name then. Also, can we set TTL to a non zero value to not mix the resolution failures with TTL expiration?

Reply:
> Ah ok, let's state this in the test name then.

Sure. I'll update the test name to reflect the same.


> Also, can we set TTL to a non zero value to not mix the resolution failures with TTL expiration?

The TTL is set to zero as the DNS name will only be removed if the TTLs of the IP addresses have expired. The reason behind this is that there might be connectivity issues due to which the DNS lookups may fail. However, as we know that the IP addresses are still valid, the resolved name for the DNS name is not removed. Thus, if the TTL is set to some non-zero value then the resolved name for the DNS name will not be removed from the status.